### ISSUE TITLE
Why the onion architecture ArchUnit test detects one violation when there are none?

### ISSUE THREAD

**Issue by Tristan107** [TNG/ArchUnit#1111]
Here is the formatted SO question :
https://stackoverflow.com/questions/76248590/why-the-onion-architecture-archunit-test-detects-one-violation-when-there-are-no

Here is a raw copy : 

I have a demo project which tries to respect strictly clean/onion/hexagonal architecture.
Here is how I configure ArchUnit test :

        @AnalyzeClasses(packages ="fr.tristan.demoassurance", importOptions = {ImportOption.DoNotIncludeTests.class, ExcludeConfig.class})
    public class ArchitectureTest {
    
        @ArchTest
        public static final ArchRule onionArchitectureRule = onionArchitecture()
            .domainModels("fr.tristan.demoassurance.domain.entity..")
            .domainModels("fr.tristan.demoassurance.domain.event..")
            .domainServices("fr.tristan.demoassurance.domain.api..")
            .applicationServices("fr.tristan.demoassurance.application.controller")
            .adapter("message", "fr.tristan.demoassurance.infrastructure.message..")
            .adapter("persistence-mongodb", "fr.tristan.demoassurance.infrastructure.repository.mongodb..")
            .adapter("persistence-mysql", "fr.tristan.demoassurance.infrastructure.repository.mysql..")
            ;
    }

ArchUnit complains about one parameter :

    was violated (1 times):
    Method <fr.tristan.demoassurance.domain.spi.message.MessageQueueSpi.send(fr.tristan.demoassurance.domain.event.PoliceAssuranceEvent)> has parameter of type <fr.tristan.demoassurance.domain.event.PoliceAssuranceEvent> in (MessageQueueSpi.java:0)

Here is the "guilty" method :

    package fr.tristan.demoassurance.domain.spi.message;
    
    import fr.tristan.demoassurance.domain.event.PoliceAssuranceEvent;
    
    public interface MessageQueueSpi {
        void send(PoliceAssuranceEvent policeAssuranceCreatedEvent);
    }

This makes no sense for me because :

1. I don't see which rule is violated, it's just not explained why a SPI interface should not have a parameter of a domain model type.

2. I have an other interface in the same `fr.tristan.demoassurance.domain.spi` parent package with a `save`method which also have a parameter `PoliceAssurance` from "domain models" and ArchUnit doesn't complain about it :

```
package fr.tristan.demoassurance.domain.spi.repository;

import java.util.List;

import fr.tristan.demoassurance.domain.entity.PoliceAssurance;
import fr.tristan.demoassurance.domain.exception.PoliceAssuranceException;

public interface PoliceAssuranceRepositorySpi {

	List<PoliceAssurance> findAll();
	PoliceAssurance save(PoliceAssurance policeAssurance) throws PoliceAssuranceException;
	PoliceAssurance findById(Integer id) throws PoliceAssuranceException;
	void deleteById(Integer id) throws PoliceAssuranceException;	
}
```

**Comment by marco-sanjuan** [comment]
Same issue here, in my test, domain ports (interfaces) methods can not have domain object as parameter.

Am I missing something?

**Comment by hankem** [comment]
`domainModels` [behaves like a setter](https://github.com/TNG/ArchUnit/blob/v1.0.1/archunit/src/main/java/com/tngtech/archunit/library/Architectures.java#L805-L828), i.e. does not extend the internal domainModelPredicate (and likewiese for `domainServices` and `applicationServices`). That is, when calling the methods multiple times, previous package identifiers are simply overwritten.

[The solution](https://stackoverflow.com/a/76252854/13134095) for @Tristan107 was to
1. pass all `domainModels` in one method call, and
2. add `mappers` and `spi` interfaces to `applicationServices` and `domainServices`, respectively.
```diff
     @ArchTest
     public static final ArchRule onionArchitectureRule = onionArchitecture()
-        .domainModels("fr.tristan.demoassurance.domain.entity..")
-        .domainModels("fr.tristan.demoassurance.domain.event..")
-        .domainServices("fr.tristan.demoassurance.domain.api..")
-        .applicationServices("fr.tristan.demoassurance.application.controller")
+        .domainModels("fr.tristan.demoassurance.domain.entity..", "fr.tristan.demoassurance.domain.event..")
+        .domainServices("fr.tristan.demoassurance.domain.api..", "fr.tristan.demoassurance.domain.spi..")
+        .applicationServices("fr.tristan.demoassurance.application.controller..", "fr.tristan.demoassurance.application.mapper..")
         .adapter("message", "fr.tristan.demoassurance.infrastructure.message..")
         .adapter("persistence-mongodb", "fr.tristan.demoassurance.infrastructure.repository.mongodb..")
         .adapter("persistence-mysql", "fr.tristan.demoassurance.infrastructure.repository.mysql..")
```

**Comment by hankem** [comment]
A discussion was started in [comments on StackOverflow](https://stackoverflow.com/a/76257355/13134095) (which is probably better continued here)
1. whether the second invocation of those methods should log a warning (or even throw an exception, but that might break intentional usages 🤔), and 
2. whether service provider interfaces should have their own category (and not be "forced" to be declared as "domain services" or "adapters").

**Comment by Tristan107** [comment]
Yes but like I said there should be more categories, like "serviceProviderInterfaces" to ensure this is actual onion architecture and not layered architecture with domain depending on persistence like in your example.
